src/proof_transcript.rs

```diff
@@ -63,41 +59,49 @@ use byteorder::{ByteOrder, LittleEndian};
 #[derive(Clone)]
 pub struct ProofTranscript {
     hash: Keccak,
+    rate: usize,
+    write_offset: usize, // index within a block where the message must be absorbed
 }
 
 impl ProofTranscript {
+    // Implementation notes
+    //
+    // The implementation has 3 layers:
+    // 1. commit/challenge - take input/output buffers <64K, responsible for disambiguation (length prefixing)
+    // 2. write/read       - take arbitrary buffers, responsible for splitting data over Keccak-f invocations and padding
+    // 3. absorb/squeeze   - actual sponge operations, outer layers ensure that absorb/squeeze do not perform unnecessary permutation
+    //
+
     /// Begin a new, empty proof transcript, using the given `label`
     /// for domain separation.
     pub fn new(label: &[u8]) -> Self {
         let mut ro = ProofTranscript {
+            // NOTE: if you change the security parameter, also change the rate below
             hash: Keccak::new_shake128(),
+            rate: 1600/8 - (2*128/8), // 168 bytes
+            write_offset: 0,
         };
+        // We will bump the version prefix each time we
+        // make a breaking change in order to disambiguate
+        // from the previous versions of this implementation.
+        ro.commit(b"ProofTranscript v2");
         ro.commit(label);
-        // makes sure the label is disambiguated from the rest of the messages.
-        ro.pad();
```

Comment:
We don't really need to pad here, as we don't pad/permute in-between multiple commitments. And `clone()` would work correctly with arbitrary offset.